nixos/lib/utils.nix

```diff
@@ -40,10 +40,39 @@ rec {
     || any (hasPrefix a'.mountPoint) b'.depends;
 
   # Escape a path according to the systemd rules, e.g. /dev/xyzzy
-  # becomes dev-xyzzy.  FIXME: slow.
-  escapeSystemdPath = s:
-   replaceChars ["/" "-" " "] ["-" "\\x2d" "\\x20"]
-   (removePrefix "/" s);
+  # becomes dev-xyzzy.
+  #
+  # NOTE: this implementation is not fully compliant with
+  # the specification in systemd.unit(5).
+  # For example, duplicate path separator characters aren't coelesced,
+  # and non-ASCII characters aren't escaped, as per `escapeSystemd`.
+  escapeSystemdPath = arg:
+    let
+      s = if builtins.isPath arg then "${arg}"
+        else if builtins.isString arg then arg
+        else throw "`escapeSystemdPath` applied to ${builtins.typeOf arg} value while a string or path was expected";
+    in
+      if s == "" || s == "/" then "-"
+      else escapeSystemd (removePrefix "/" (removeSuffix "/" s));
+
+  # Escape a string according to the rules described in systemd.unit(5).
+  # e.g. "My Container 1" becomes "My\x20Container\x201".
+  #
+  # NOTE: this implementation is not fully compliant with systemd.
+  # Many characters which aren't ASCII alphanumerics are incorrectly
+  # passed through unescaped.
+  escapeSystemd = str: let
```

Comment:
Given that `escapeSystemdExecArg` also lives in this file, this function name reads too generic to me unless the former will also make use of it.